dagql/server.go

```diff
@@ -0,0 +1,890 @@
+				return fmt.Errorf("%s: %w", sel.Name(), err)
+			}
+			results.Store(sel.Name(), res)
+			return nil
+		})
+	}
+	if err := pool.Wait(); err != nil {
+		return nil, err
+	}
+
+	resultsMap := make(map[string]any)
+	results.Range(func(key, value any) bool {
+		resultsMap[key.(string)] = value
+		return true
+	})
+	return resultsMap, nil
+}
+
+// Load loads the object with the given ID.
+func (s *Server) Load(ctx context.Context, id *idproto.ID) (Object, error) {
+	var base Object
+	var err error
+	if id.Parent != nil {
+		base, err = s.Load(ctx, id.Parent)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		base = s.root
+	}
+	astField := &ast.Field{
+		Name: id.Field,
+	}
+	vars := map[string]any{}
+	for _, arg := range id.Args {
+		vars[arg.Name] = arg.Value.ToInput()
+		astField.Arguments = append(astField.Arguments, &ast.Argument{
+			Name: arg.Name,
+			Value: &ast.Value{
+				Kind: ast.Variable,
+				Raw:  arg.Name,
+			},
+		})
+	}
+	sel, _, err := base.ObjectType().ParseField(ctx, astField, vars)
+	if err != nil {
+		return nil, fmt.Errorf("parse field %q: %w", astField.Name, err)
+	}
+	sel.Nth = int(id.Nth)
+	res, id, err := s.cachedSelect(ctx, base, sel)
+	if err != nil {
+		return nil, err
+	}
+	return s.toSelectable(id, res)
+}
+
+func (s *Server) Select(ctx context.Context, self Object, dest any, sels ...Selector) error {
+	for _, sel := range sels {
+		res, id, err := s.cachedSelect(ctx, self, sel)
+		if err != nil {
+			return err
+		}
+		self, err = s.toSelectable(id, res)
+		if err != nil {
+			return err
+		}
+	}
+	return assign(reflect.ValueOf(dest).Elem(), self)
+}
+
+func LoadIDs[T Typed](ctx context.Context, srv *Server, ids []ID[T]) ([]T, error) {
+	out := make([]T, len(ids))
+	eg := new(errgroup.Group)
+	for i, id := range ids {
+		i := i
+		id := id
+		eg.Go(func() error {
+			val, err := id.Load(ctx, srv)
+			if err != nil {
+				return err
+			}
+			out[i] = val.Self
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+type idCtx struct{}
+
+func idToContext(ctx context.Context, id *idproto.ID) context.Context {
+	return context.WithValue(ctx, idCtx{}, id)
+}
+
+func CurrentID(ctx context.Context) *idproto.ID {
+	val := ctx.Value(idCtx{})
+	if val == nil {
+		return nil
+	}
+	return val.(*idproto.ID)
+}
+
+func (s *Server) cachedSelect(ctx context.Context, self Object, sel Selector) (res Typed, chained *idproto.ID, rerr error) {
+	chainedID, err := self.IDFor(ctx, sel)
+	if err != nil {
+		return nil, nil, err
+	}
+	ctx = idToContext(ctx, chainedID)
+	dig, err := chainedID.Canonical().Digest()
+	if err != nil {
+		return nil, nil, err
+	}
+	doSelect := func(ctx context.Context) (Typed, error) {
+		return self.Select(ctx, sel)
+	}
+	if s.telemetry != nil {
+		doSelect = s.telemetry(ctx, self, chainedID, doSelect)
+	}
+	var val Typed
+	if chainedID.IsTainted() {
+		val, err = doSelect(ctx)
+	} else {
+		val, err = s.Cache.GetOrInitialize(ctx, dig, doSelect)
```

Comment:
There's some low-hanging-fruit optimizations possible here, e.g. when we're loading an ID we could do a direct cache lookup for the final ID instead of re-evaluating it and going through N cache hits for the depth of the ID. Just kept it simple for now.